logbook-api/src/main/java/org/zalando/logbook/HttpHeaders.java

```diff
@@ -0,0 +1,137 @@
+package org.zalando.logbook;
+
+import javax.annotation.CheckReturnValue;
+import javax.annotation.concurrent.Immutable;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.BiPredicate;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.UnaryOperator;
+
+import static org.zalando.logbook.DefaultHttpHeaders.EMPTY;
+
+/**
+ * An immutable multi-map representing HTTP headers. It offers three kinds of
+ * operations:
+ *
+ * {@code update} is essentially {@link Map#put(Object, Object)}, but returns
+ * a new instance with the updated entries.
+ *
+ * {@code delete} is essentially {@link Map#remove(Object)}, but returns a new
+ * instance without the deleted headers.
+ *
+ * {@code apply} applies a function to all entries or a subset of them and
+ * applies the changes. If the given operator returns a null value the entry
+ * will be deleted otherwise updated.
+ */
+@Immutable
+public interface HttpHeaders extends Map<String, List<String>> {
+
+    @CheckReturnValue
+    HttpHeaders update(String name, String... values);
+
+    @CheckReturnValue
+    HttpHeaders update(String name, Collection<String> value);
+
+    @CheckReturnValue
+    HttpHeaders update(Map<String, List<String>> headers);
+
+    @CheckReturnValue
+    HttpHeaders apply(String name, UnaryOperator<List<String>> operator);
+
+    @CheckReturnValue
+    HttpHeaders apply(
+            Collection<String> names,
+            BiFunction<String, List<String>, Collection<String>> operator);
+
+    @CheckReturnValue
+    HttpHeaders apply(
+            BiPredicate<String, List<String>> predicate,
+            BiFunction<String, List<String>, Collection<String>> operator);
+
+    @CheckReturnValue
+    HttpHeaders apply(
+            BiFunction<String, List<String>, Collection<String>> operator);
+
+    @CheckReturnValue
+    HttpHeaders delete(String... names);
+
+    @CheckReturnValue
+    HttpHeaders delete(Collection<String> names);
+
+    @CheckReturnValue
+    HttpHeaders delete(BiPredicate<String, List<String>> predicate);
+
+    static HttpHeaders empty() {
+        return EMPTY;
+    }
+
+    static HttpHeaders of(final Map<String, List<String>> headers) {
+        return empty().update(headers);
+    }
+
+    static <T, U> BiPredicate<T, U> predicate(final Predicate<T> predicate) {
+        return (t, u) -> predicate.test(t);
+    }
+
+    // deprecated stuff from here on till the end
```

Comment:
worth to keep them around at all?

Reply:
Just to mark them as deprecated so that IDEs are highlighting our immutability aspect.